packages/nimble-components/src/table/specs/table-row-grouping-hld.md

```diff
@@ -0,0 +1,238 @@
+
+```
+export class TableColumnTextGroupHeaderView extends TableGroupHeaderView<
+TableColumnTextColumnConfig
+> {
+    @observable
+    public override columnConfig!: TableColumnTextColumnConfig;
+
+    @volatile
+    public get content(): string {
+        return typeof this.groupHeaderValue === 'string'
+            ? this.groupHeaderValue
+            : this.columnConfig.placeholder;
+    }
+}
+
+const tableColumnTextGroupHeaderView = TableColumnTextGroupHeaderView.compose({
+    baseName: 'table-column-text-group-header',
+    template: html<TableColumnTextGroupHeaderView>`
+        <nimble-text-field readonly="true" value="${x => x.groupHeaderValue}" placeholder="${x => x.columnConfig.placeholder}">
+        </nimble-text-field>`,
+    styles: /* styling */
+});
+DesignSystem.getOrCreate().withPrefix('nimble').register(tableColumnTextGroupHeaderView());
+```
+
+### Rendering the row group
+
+Rendering a row group has concerns beyond just rendering the grouping value. It also has a button for controlling the expand/collapse state of a row group (indented for sub-groups) as well as potentially rendering auxiliary information such as the total number of items within a group. We can provide a new `TableGroupRow` element to manage all of these:
+
+```
+export class TableGroupRow extends FoundationElement {
+    @observable
+    public groupRowValue?: TableFieldValue;
+
+    @observable
+    public nestingLevel: number = 0;
+
+    @observable
+    public leafItemCount?: number;
+
+    @observable
+    public columnConfig?: unknown;
+
+    @observable
+    public isExpanded?: boolean;
+
+    @observable
+    public groupRowHeaderViewTag?: string;
+
+    public onExpandGroupToggle(): void {
+        const detail: TableGroupExpandToggleEventDetail = {
+        };
+        this.$emit('group-expand-toggle', detail);
+    }
+
+    ...
+}
+
+const tableGroupRowElement = TableGroupRow.compose({
+    baseName: 'table-group-row',
+    template: html<TableGroupRow>`
+    <template
+        role="rowgroup"
+        style="--ni-private-row-group-indent-level: ${x => x.nestingLevel};"
+    >
+        <nimble-button
+            appearance="ghost"
+            content-hidden
+            @click=${x => x.onExpandGroupToggle()}>
+            ${when(x => x.isExpanded, html`
+                <nimble-icon-arrow-expander-down slot="start"></nimble-icon-arrow-expander-down>
+            `)}
+            ${when(x => !x.isExpanded, html`
+                <nimble-icon-arrow-expander-right slot="start"></nimble-icon-arrow-expander-right>
+            `)}
+        </nimble-button>
+        <${x => x.rowGroupState.groupRowHeaderViewTag}
+            :rowGroupValue="${x => x.groupRowValue}"
+            :columnConfig="${x => x.columnConfig}"
+            >
+        </{x => x.rowGroupState.groupRowHeaderViewTag}> (${x => x.leafItemCount})
```

Comment:
I had to change the implementation to the following to get it to work, which mirrors what @msmithNI did in his prototype:

```html
  ${when(x => x.groupHeaderViewTag, x => {
  const groupHeaderViewTag = x.groupHeaderViewTag!;
  return html<TableGroupRow>`
          <${groupHeaderViewTag}
              :groupHeaderValue="${y => y.groupRowValue}"
              :columnConfig="${y => y.columnConfig}"
              class="group-header-value"
              >
          </${groupHeaderViewTag}>
      `;
  })}
```

I'll update the HLD accordingly.

Reply:
We need to cache that result or something. As-is that is creating a new instance during each evaluation of template binding which is a lot of unnecessary template string parsing work.

It could probably be done by in the class having something like:

```ts
@observable
public groupHeaderViewTemplate?;

@observable
public groupHeaderViewTag?;

groupHeaderViewTagChanged () {
    this.groupHeaderViewTemplate = html<TableGroupRow>`
          <${this.groupHeaderViewTag}
              :groupHeaderValue="${y => y.groupRowValue}"
              :columnConfig="${y => y.columnConfig}"
              class="group-header-value"
              >
          </${this.groupHeaderViewTag}>
    `;
}

```

And then in the template just use that instance:
```ts
${when(x => x.groupHeaderViewTag, x.groupHeaderViewTemplate)}
```

i.e. Check the `Example: Dynamic Composition` in the [template composition docs](https://www.fast.design/docs/fast-element/using-directives#composing-templates).

Ideally there would be a single instance shared across all TableGroupRow instances that are using it.

Again maybe it's implementation details too specific for the spec but up to y'all how / when to discuss it.

Reply:
I'm going to consider this comment to be implementation detail that lies outside of what this spec needs to resolve, and will follow-up with it in the actual implementation.